hq_superset/utils.py

```diff
@@ -120,16 +129,57 @@ class DomainSyncUtil:
     def __init__(self, security_manager):
         self.sm = security_manager
 
-    def _ensure_domain_role_created(self, domain):
-        # This inbuilt method creates only if the role doesn't exist.
-        return self.sm.add_role(get_role_name_for_domain(domain))
+    def sync_domain_role(self, domain):
+        """
+        This method ensures the roles are set up correctly for a particular domain user.
 
-    def _ensure_schema_perm_created(self, domain):
-        menu_name = self.sm.get_schema_perm(get_hq_database(), get_schema_name_for_domain(domain))
-        permission = self.sm.find_permission_view_menu("schema_access", menu_name)
-        if not permission:
-            permission = self.sm.add_permission_view_menu("schema_access", menu_name)
-        return permission
+        The user gets assigned at least 3 roles in order to function on any domain:
+        1. hq_user_role: gives access to superset platform
+        2. domain_schema_role: restricts user access to specific domain schema
+        3. domain_user_role: restricts access for particular user on domain in accordance with how the permissions
+        are defined on CommCare HQ.
+
+        Any additional roles defined on CommCare HQ will also be assigned to the user.
+        """
+        hq_user_role = self._ensure_hq_user_role()
+        domain_schema_role = self._create_domain_role(domain)
+
+        additional_roles = self._get_additional_user_roles(domain)
+        if not additional_roles:
+            return False
+
+        current_user.roles = [hq_user_role, domain_schema_role] + additional_roles
+
+        self.sm.get_session.add(current_user)
+        self.sm.get_session.commit()
+        return True
+
+    def _ensure_hq_user_role(self):
+        """
+        This role is the bare minimum required for a user to be able to have an account on
+        superset
+        """
+        hq_user_role = self.sm.add_role(HQ_USER_ROLE_NAME)
+
+        hq_user_base_permissions = [
+            self.sm.add_permission_view_menu("can_profile", "Superset"),
+            self.sm.add_permission_view_menu("can_recent_activity", "Log"),
+        ]
+        self.sm.set_role_permissions(hq_user_role, hq_user_base_permissions)
+
+        if hq_user_role not in current_user.roles:
```

Comment:
nit: should this check be on the top so the role isn't set if its already present?